src/cargo/sources/git/utils.rs

```diff
@@ -733,3 +733,8 @@ fn github_up_to_date(handle: &mut Easy, url: &Url, oid: &git2::Oid) -> bool {
 
     try!(handle.response_code().ok()) == 304
 }
+
+#[inline]
+pub fn is_local_filesystem(url: &Url) -> bool {
+    url.scheme() == "file"
```

Comment:
Hm could we avoid this special casing? I don't think this is actually used that much in practice and treating all URLs uniformly would make writing tests much easier I think.

Reply:
Thanks for the review. I will improve the tests.

But i am not sure about treat all URLs uniformly.

As far, as i understand:

Now we did not distinguish between a local or network git repository.
But in offline mode, we would like to do this. We can allow to work with local repositories, but forbid to work with network repositories.

This can be useful for usual offline cases. For example, we have many projects A, B ... Z,that depends on the 'master' branch of our library LIB located in local filesystem.
We want to add some functionality to LIB without breaking many projects, and then use that functionality in some of our projects. We do not have network.

We do this in separate branch (for example, 'my_cool_feature'), as usually.
But after merge 'my_cool_feature' to 'master' branch of the LIB, projects A ... Z can not see top of the master.
There is no way to cache top of the master to the cargo cache without networking.
This will look strange: there is a local git repository - but we can not use it without network.

This code (and some other places in pr) is needed to handle this case.
If we remove this, the code and tests will be easier, but we will not be able to do this.
If we remove the pull-up from the local git repository into the cache, this will worsen usability.

It turns out that if this functionality is missing, we can develop binaries - but we can not use new changes in libraries in the offline mode.

Perhaps this case requires additional discussion?

Reply:
Oh yeah for sure we *could* make this distinction but I've never in practice seen anyone using a local git repository. Instead these are just used in tests, where it's quite useful to use the local filesystem as if it were the network to be able to write tests against.

Perhaps let's not make a special distinction here and leave that to a future PR?